src/synthcity/plugins/core/models/tabular_ddpm/modules.py

```diff
@@ -0,0 +1,515 @@
+import torch.nn.functional as F
+import torch.optim
+from torch import Tensor
+
+ModuleType = Union[str, Callable[..., nn.Module]]
+
+
+class SiLU(nn.Module):
+    def forward(self, x: Tensor) -> Tensor:
+        return x * torch.sigmoid(x)
+
+
+def timestep_embedding(timesteps: Tensor, dim: int, max_period: int = 10000) -> Tensor:
+    """
+    Create sinusoidal timestep embeddings.
+
+    :param timesteps: a 1-D Tensor of N indices, one per batch element.
+                      These may be fractional.
+    :param dim: the dimension of the output.
+    :param max_period: controls the minimum frequency of the embeddings.
+    :return: an [N x dim] Tensor of positional embeddings.
+    """
+    half = dim // 2
+    freqs = torch.exp(
+        -math.log(max_period)
+        * torch.arange(start=0, end=half, dtype=torch.float32)
+        / half
+    ).to(device=timesteps.device)
+    args = timesteps[:, None].float() * freqs[None]
+    embedding = torch.cat([torch.cos(args), torch.sin(args)], dim=-1)
+    if dim % 2:
+        embedding = torch.cat([embedding, torch.zeros_like(embedding[:, :1])], dim=-1)
+    return embedding
+
+
+def reglu(x: Tensor) -> Tensor:
+    """The ReGLU activation function from [1].
+    References:
+        [1] Noam Shazeer, "GLU Variants Improve Transformer", 2020
+    """
+    if not (x.shape[-1] % 2 == 0):
+        raise AssertionError
+    a, b = x.chunk(2, dim=-1)
+    return a * F.relu(b)
+
+
+def geglu(x: Tensor) -> Tensor:
+    """The GEGLU activation function from [1].
+    References:
+        [1] Noam Shazeer, "GLU Variants Improve Transformer", 2020
+    """
+    if not (x.shape[-1] % 2 == 0):
+        raise AssertionError
+    a, b = x.chunk(2, dim=-1)
+    return a * F.gelu(b)
+
+
+class ReGLU(nn.Module):
+    """The ReGLU activation function from [shazeer2020glu].
+
+    Examples:
+        .. testcode::
+
+            module = ReGLU()
+            x = torch.randn(3, 4)
+            assert module(x).shape == (3, 2)
+
+    References:
+        * [shazeer2020glu] Noam Shazeer, "GLU Variants Improve Transformer", 2020
+    """
+
+    def forward(self, x: Tensor) -> Tensor:
+        return reglu(x)
+
+
+class GEGLU(nn.Module):
+    """The GEGLU activation function from [shazeer2020glu].
+
+    Examples:
+        .. testcode::
+
+            module = GEGLU()
+            x = torch.randn(3, 4)
+            assert module(x).shape == (3, 2)
+
+    References:
+        * [shazeer2020glu] Noam Shazeer, "GLU Variants Improve Transformer", 2020
+    """
+
+    def forward(self, x: Tensor) -> Tensor:
+        return geglu(x)
+
+
+def _make_nn_module(module_type: ModuleType, *args: Any) -> nn.Module:
+    return (
+        (
+            ReGLU()
+            if module_type == "ReGLU"
+            else GEGLU()
+            if module_type == "GEGLU"
+            else getattr(nn, module_type)(*args)
+        )
+        if isinstance(module_type, str)
+        else module_type(*args)
+    )
+
+
+class MLP(nn.Module):
```

Comment:
There is another MLP in the codebase, at synthcity.plugins.core.models.mlp

Consider reusing that class instead, and drop this one